### ISSUE TITLE
[LLVM][Feature] Kernel specialization

### ISSUE THREAD

**Issue by mciprian13** [pytorch/glow#4484]
**Usecase/Assumptions**
- Assume I have a 3rd party kernel library (with e.g. convolution, gemm, etc) denoted from now on with **LIBX** designed and optimized for a particular target (e.g. ARM) and I want to integrate it in Glow as easy/quickly as possible without even creating a new Backend.
- The target for which **LIBX** is developed is assumed to have LLVM support and therefore the **LIBX** specialization is assumed to go through LLVMIRGen.
- **LIBX** is assumed to have only a small subset of operators and so cannot replace **LIBJIT** entirely.
- The intent is to have a logic which, upon user request, tries to specialize some of the IR instructions to kernel implementations from **LIBX** instead of **LIBJIT**.
- There might be also constraints associated to the **LIBX** kernels, for example:
  - **libx_conv2d_f** - might not have support for dilation or could only have support for square padding/stride, etc
  - **libx_conv2d_nchw_f** – has support only for NCHW layout
  - **libx_fully_connected_f** - requires the constant weights to be transposed or arranged in a very specific manner
  - **libx_mat_mul_f** - requires some scratch memory
  - **libx_mat_mul_i8** – is implemented for symmetric quantization schema
- The integration code must express all the conditions under which the mapping can be done: if all the kernel constraints are met then the mapping is done, otherwise the mapping falls back to the **LIBJIT** kernel implementations. 

**Requirements**
- Design infrastructure for integrating optimized libraries in the LLVMIRGen workflow.
- The infrastructure should promote code isolation similar to the Backend registration in order for Glow users to integrate their own libraries in a private manner, without having to deal with conflicts from upstream.
- The infrastructure should be part of the CPU/LLVM backend and should allow easy “registration” of new libraries. This mechanism (Kernel/Library Specializer) could be considered something of a new hierarchy (similar to backend) below the CPU/LLVM backend.
- New libraries will be contained in separate folders, e.g. parallel with the “libjit” folder. Each such folder will contain:
  - Logic for expressing constraints for each kernel in terms of:
    - Target type: if kernel code contains assembly for e.g. ARM Cortex-M then the kernel specialization kicks in only if the compilation target is ARM Cortex-M.
    - Data type, layout, quantization parameters
    - Operands sizes
    - Instruction parameters
    - Others …
  - Mechanism to request scratch allocation before LLVMIRGen
  - Mechanism to rearrange/redefine its constant weights buffers before LLVMIRGen.
  - LLVMIRGen code for the kernels
  - It would be nice if the Kernel Specializer could have tentacles/hooks upper in the compile flow such that it could modify the graph IR (e.g. add a transpose node before/after the convolution in order to allow a NCHW convolution, etc) such that where constraints are not met it could enforce them if possible.
- The kernel specialization is performed per user request, using maybe a command line parameter for compilation like `-use-libx`
- Multiple kernel specializers can be requested, e.g. `-use-libx -use-liby` such that if for a given instruction **LIBX** specialization failed then **LIBY** specialization is tried. If **LIBY** specialization also failed, then **LIBJIT** will be used for that instruction.

**Comment by mciprian13** [comment]
@opti-mix @jfix71 I raised this issue to have more like a discussion. This is something I need for a long time now because 90% of my work with Glow is to integrate optimized kernels done by others (3rd party). The quantity of code I write for this currently is more than I can handle and I need something to speed things up. In my opinion this would greatly benefit the entire Glow community.
I would gladly work on this but I need to have a discussion with you and maybe for you to draw me some guidelines.
What are your thoughts on this?
Thanks!

**Comment by jfix71** [comment]
Hi @mciprian13 -- so I have less familiarity with LLVM-based backends than @opti-mix, he would have better suggestions specific to the LLVM backend. But from a pure Glow perspective, one way this could be done is to use backend-specific Instructions and `Backend::generateInst()`. You could have e.g. `YourBackendMatMulInst` that matches the Glow default `MatMulInst`. You could add an input like `Scratch` to it too (see the `TopKInst`) if you know you may need it as you mentioned for `libx_mat_mul_f `, and also a member for what specific kernel you want to call, e.g. `addMember(MemberType::String, "KernelName")`. Then in your backend's `generateInst()` you could look for cl options for what sorts of libs to try to use, and then just add rules for finding/replacing the default Glow inst with `YourBackendConvolutionInst` and the desired kernel. 

If you wanted it to be even more generic you could make a single generic backend-specific Inst to swap in instead of adding Insts for each one you may want to use a specialized kernel. See here for what we do for NNPI specialized DSP kernels (this is at the Node level, but something similar could happen at the Inst level too). https://github.com/pytorch/glow/blob/644d26f58c3dbe56e5d9b7ccc50031a01c300f3b/lib/Backends/NNPI/ClassGen/NNPISpecificNodes.h#L18-L27

Would these directions would help solve your problem?

**Comment by opti-mix** [comment]
@mciprian13 I understand the motivation for your request. And I agree that it is somewhat painful to integrate 3rd party specialized libraries into Glow's CPU backend or other Glow backends. 

Ideally, the mechanisms you describe should be reusable to a big extent for other backends. In case of backends for HW accelerators, you also often have optimized custom kernels available as libraries or even implemented in HW in addition to the set of kernels provided/generated by Glow and Glow compiler also needs to pick the best one. How this specific backend is implemented, i.e. based on LLVM or not, is a relatively minor piece of this big picture.

@jfix71 made good points about how you could use existing mechanisms to help you achieving what you need. TensorLayouts could probably help to automatically transform the graph and bring layouts into the required form. But as far as I can tell from your request, all this is by far not enough for you.

It sounds like you really want to have general mechanisms that would involve some pattern matching, kernel selection logic and graph/instructions transformation logic to pick the right kernel candidate and then prepare its invocation. While I agree that is would be nice to have this, it sounds like a pretty significant amount of work across different IRs and abstraction levels required to support this. Many other ML frameworks (and domain-specific compilers in general) have faced a similar issue, I'd say. And I don't remember any of them having solved them in a nice and clean way not involving a lot of manual tweaking and hard-wiring some of the library specific logic in the compiler to select kernels. I suspect this is for a good reason, as this is very non-trivial! If I'm wrong, could you please refer me an existing system that nicely provides the functionality you ask for?

Basically, what I'm trying to say is: This feature would likely need a well-thought design and a lot of development effort, unless we come up with something small (in terms of development) and very smart. Thus, the benefits should be compared against the development costs for this feature. We should also discuss, if we e.g. could achieve 80% of the benefits with 20% of efforts.

**Comment by mciprian13** [comment]
@jfix71 Currently we integrate the kernels in a nasty way (don't judge me 😄 cuz this path was chosen when I did not have experience in Glow):
- we don't have our own backend ... we just hijacked the CPUBackend
- in TransformPostLowering we conditionally specialize generic nodes to CPU specific nodes (e.g. XConvolutionNode) which has some extra members (KernelName, ScratchSize)
- we do the IRGen to CPU specific instructions which can have an extra scratch operand so we do it manually in order to allocate the scratch (**currently this requires a lot of dummy/boilerplate coding which I hate**)
- the LLVMIRGen we do it manually of course and I don't expect it to be otherwise
- the library source code is compiled to LLVM-IR and appended to libjit
- this kind of stuff we do which needs to remain private has of course poor to NO isolation from the rest of the CPU Backend code so we always expect some conflicts

I will have a look on the things you mentioned (`Backend::generateInst()` and the NNPI generic example) since I'm not familiar with them.

@opti-mix I don't want anything magic like you suggested (automated pattern matching, etc) but just some means to make the integration faster/easier and more reliable. I like the idea of the 20% effort for 80% benefit. I propose to go incremental with small useful features.

Here's one simple feature I would benefit from (to write less code): automatic IRGen between a Node and an Instruction which has (relative to the Node) one extra output operand - `Scratch`. One proposal would be to have in the `InstrBuilder` an option called `requireScratch` which if used it automatically:
- Creates a new output operand for the instruction called `Scratch`
- Declares a new method for the instruction called `getScratchSize()` which must be manually defined (in-place or somewhere else) to compute the scratch size in bytes based on the instruction parameters
- The scratch allocation must also allow for scratch size 0 since in some cases the scratch size is a dynamic function of the input/output sizes and in some particular cases it can be 0 ... I think @jfix71 made an update to allow tensors with size 0 ...
- It can also build the IRGen automatically such that the Scratch operand is allocated with the right size
- Although this is an egg-chicken problem since in IRGen I need to call the method 'getScratchSize()` to allocate the scratch operand before having the Instruction object ... 
Do you have a solution for this (first) problem?
Thanks!

**Comment by ayermolo** [comment]
Just throwing it out there as a potential short term solution. Maybe it's time to create your own backend, and inherit from CPU backend. In legalizer if layer needs additional things done you can do it there, or if additional parameters are required replace it with backend specific OP. During LLVM IRGen you can fall back to CPU LLVMIRGen for nodes that you just want libjit to handle.

**Comment by mciprian13** [comment]
@opti-mix @ayermolo I agree it is time for me to create a Backend. I want to fork a Backend from the LLVM backend similar to how the CPU does. In order to do this basically I'll start by copying the code of the CPU Backend and rename stuff. There are two things at this point I'm not happy with:
1. The source file `CPUBackend.cpp` contains logic which in my opinion should be moved in the LLVM Backend. Shouldn't all the logic within the function `isOpSupported` be moved to the LLVMBackend with the exception of the `CPUMaxSplat` and `CPUConvDKKC8` which are the only nodes specific to the CPU backend ??? Because the LLVM backend should be responsible with handling all the common nodes and the CPU should be responsible with handling only the specific nodes right?
2. Currently there is no separation between the source code of the LIBJIT used by LLVM backend and the specific part used by CPU backend. For example in my case I would want to put all my kernels in a separate directory (say `libjit_for_my_backend`) and then have some CMake infrastructure to register my kernel folder and then compile and link the source code to the big `libjit`. For the CPU backend in my opinion it would be cleaner to have a separate folder `libjit_cpu` where the implementation for `CPUMaxSplat` and `CPUConvDKKC8` should be.
WDYT??

**Comment by opti-mix** [comment]
@mciprian13 

Good points!

Regarding (1): I see what you mean and it makes sense, in general. At the same time, based on my experience, new LLVMBackend-based backends often initially support only a small subset of operations. Which means they need to heavily customize the logic for isOpSupported. Over time, as they support more and more operations, they become more and more similar to the standard handling. Another axis how they different from the current CPU backend is the set of payload datatypes they support. Some of the backends may support only quantized payloads or only FP32 or only FP16 payloads, because these are the only data types supported by their HW. Thus if we move isOpSupported into a base class, we probably need to provide some mechanisms to specify which payload types are supported (e.g. which FP types, which integer types) and use those in all the checks. But maybe there is another way to handle this issue? WDYT?

Regarding (2): I agree we need a better way for having a "composite" libjit, consisting of multiple parts. More over, IIRC, there were even some requests for a single backend to provide different versions of BC for different target architectures/HW models.

**Comment by mciprian13** [comment]
@opti-mix Sorry for getting back to this after so long.

Regarding (1): I only advocate for moving the common part of the isOpSupported() from CPUBackend to LLVMBackend such that new LLVM based backends which want to reuse all the functionality but only specialize a couple of operators (like our private backend) do not have to rewrite/duplicate all that code. Other backends are free NOT to use the isOpSupported() from the LLVM backend. I opened a PR #4621 for this. Can you take a look?

Regarding (2): I think I will refactor the CMake infrastructure around LIBJIT to allow current/new LLVM based backends to build more easily a backend specific libjit with some options like linking multiple libjits for a composite one, etc

**Comment by mciprian13** [comment]
@opti-mix Regarding (2) I also left a PR here: #4627. Please have a look.
